Fixed browser history for profile tabs

rhysyngsun opened the pull request:
#### What are the relevant tickets?

Fixes #1363 

#### What's this PR do?

Reintroduces named urls for the personal/education/professional tabs on the profile setup.

#### Where should the reviewer start?

Take a look at `PersonalTab` (other tabs are similar) and the routes in `dashboard_routes.js`. Also look at `currentOrFirstIncompleteStep()` in `util/util.js`.

#### How should this be manually tested?

Sign up with a new account, verify you're directed to /profile/personal by default and that you're redirected there, particularly if trying to reach one of the later steps (e.g. /profile/education or /profile/professional). You shouldn't be allowed into those later steps until you've completed the personal tab.

#### Any background context you want to provide?

We previously had named urls for each of these tabs, this PR reintroduces them, but also addresses the [original issue](https://github.com/mitodl/micromasters/issues/627) for the [PR that removed them](https://github.com/mitodl/micromasters/pull/657) so we don't have a regression on that.

#### What GIF best describes this PR or how it makes you feel?

![Moonwalk](https://cloud.githubusercontent.com/assets/28598/20491148/76023bda-afde-11e6-8fc9-5bd4280395d5.gif)

aliceriot commented:
functionality looks good! reading code now.

aliceriot commented on static/js/containers/ProfilePage.js:
```diff
@@ -2,105 +2,95 @@
+        // or the least complete step below the current one
+        let idealStep = currentOrFirstIncompleteStep(profileStep, step);
+        if (profileStep !== idealStep) {
+          // if learner is on a step they shouldn't be on yet, redirect them to the correct one
```
I don't think these comments are super necessary, you can keep them if you want though.

aliceriot replied:
(b/c the code is pretty clear here)

aliceriot commented on static/js/containers/ProfilePage_test.js:
```diff
@@ -1,4 +1,6 @@
 /* global SETTINGS: false */
+import '../global_init';
```
whats the reason for this import?

rhysyngsun replied:
I swear there was a way I was running the tests where it failed without that import, but I can't reproduce it or remember exactly what  the error was. I'll remove it.

noisecapella replied:
Maybe you were running the tests through your IDE? We explicitly include this file in our mocha command in `package.json`, maybe you need to edit your test configuration to do the same for your IDE

aliceriot commented on static/js/containers/ProfilePage_test.js:
```diff
@@ -127,35 +126,31 @@ describe("ProfilePage", function() {
         let updatedProfile = Object.assign({}, USER_PROFILE_RESPONSE, {
           filled_out: filledOutValue,
-          education: []
+          e3ducation: []
```
typo?

rhysyngsun commented on static/js/util/util.js:
```diff
@@ -180,6 +181,31 @@ export function makeProfileProgressDisplay(active: string) {
+  const idealStepIndex = Math.min(stepIndex, currentStepIndex);
+
+  return PROFILE_STEP_ORDER[idealStepIndex];
+}
```
Good idea, still need to manually test, but this handles the nulls and passes my unit tests:

```es6
const getStepIndices = R.map(R.indexOf(R.__, PROFILE_STEP_ORDER));
const lastStep = R.last(PROFILE_STEP_ORDER);

export const currentOrFirstIncompleteStep = R.compose(
    R.nth(R.__, PROFILE_STEP_ORDER),
    R.reduce(R.min, Infinity),
    getStepIndices,
    R.append(lastStep),
    R.reject(R.isNil),
    R.pair
);
```

rhysyngsun commented on static/js/containers/ProfilePage.js:
```diff
@@ -2,105 +2,95 @@
+  renderContent(username: string, profileInfo: ProfileGetResult, profile: Profile, currentStep: ?string) {
+    if (R.isNil(profileInfo)) {
+      return '';
+    }
```
In practice, at least the first time `ProfilePage`'s render method gets hit, `profileInfo` is `undefined`.

aliceriot replied:
Oh that makes sense, because the dashboard API response hasn't come back yet.

rhysyngsun commented:
Hmm, I didn't think this was actually a coverage decrease...looking into that.

aliceriot commented:
would you mind adding new commits to the PR when you update it? It makes the review a little bit easier and makes it easier to follow your progress.

rhysyngsun commented:
@aliceriot gah, sorry! noticed your comment too late. I'm an obsessive rebaser/squasher.

To make up for that, here's a list of change in my last push:

- Removed extraneous `if (profile !== undefined){}` wrapping lines 40-44 here: https://github.com/mitodl/micromasters/pull/1814/files#diff-7070f6579a91d7efe615fe81a22bdafeR40
- Fixed the typo
- Refactored `currentOrFirstIncompleteStep` per your recommendations here: https://github.com/mitodl/micromasters/pull/1814/files#diff-3c2e3ec6197c815c23dad44b122f3002R184
- Removed `nextStep` / `prevStep` from `SettingsPage` since they're no longer used and it was indirectly reducing test coverage: https://github.com/mitodl/micromasters/pull/1814/files#diff-0edec7f878bf81603aeb87f756925f46R26
- Added these tests to boost `ProfilePage` test coverage back to 100%: https://github.com/mitodl/micromasters/pull/1814/files#diff-c2220af76fb0be7547e07aabb356fe9eR136
- Renamed a test constant and added another below it: https://github.com/mitodl/micromasters/pull/1814/files#diff-c2220af76fb0be7547e07aabb356fe9eR50

aliceriot commented:
It's fine to rebase, just save the squash for merge-time.

aliceriot commented:
I rebase every time I update my branches on github too.

aliceriot commented on static/js/components/PersonalTab_test.js:
```diff
@@ -25,6 +33,13 @@ describe("PersonalTab", () => {
+    if ( window.localStorage === undefined ) {
+      window.localStorage = localStorageMock();
+    }
+    store = configureMainTestStore(rootReducer);
```
You could use an `integrationTestHelper` here to get this for free - might be worth it even if you're not going to use it's various test methods (`listenForActions` and so on).

rhysyngsun replied:
done

aliceriot commented on static/js/containers/ProfilePage_test.js:
```diff
@@ -127,35 +129,59 @@ describe("ProfilePage", function() {
+
+  describe('profile completeness', () => {
+    let checkStep = () => helper.store.getState().ui.profileStep;
+
```
could you get rid of this extra line here?

rhysyngsun replied:
Yup will do, also just caught that `checkStep` is now defined twice in this file and is also a bit of a misnomer.

aliceriot replied:
haha, good catch

aliceriot commented on static/js/components/ProfileProgressControls.js:
```diff
@@ -32,22 +37,23 @@ export default class ProfileProgressControls extends React.Component {
     }
-    if(nextStep) {
+    if (nextUrl) {
+
```
blank line

aliceriot commented on static/js/constants.js:
```diff
@@ -31,6 +32,11 @@ export const PROFILE_STEP_LABELS = new Map([
   [EDUCATION_STEP, "Education"],
   [EMPLOYMENT_STEP, "Professional"]
 ]);
```
could you add a blank line here?

aliceriot commented on static/js/containers/App_test.js:
```diff
@@ -80,48 +81,49 @@ describe('App', () => {
     let checkStep = () => helper.store.getState().ui.profileStep;
 
-    it('redirects to /profile if profile is not complete', () => {
+
```
blank line

aliceriot commented on static/js/containers/ProfilePage_test.js:
```diff
@@ -127,35 +131,56 @@ describe("ProfilePage", function() {
+  });
+
+  describe('profile completeness', () => {
+
```
blank line
